crates/red_knot/src/types.rs

```diff
@@ -37,40 +38,137 @@ impl TypeEnvironment {
     }
 }
 
+#[derive(Debug)]
 pub(crate) enum Type {
-    Class(ClassType),
+    // a specific function
     Function(FunctionType),
+    // the set of Python objects with this class in their MRO
+    Instance(ClassType),
+    Union(UnionType),
+    Intersection(IntersectionType),
+    // the dynamic or gradual type: a statically-unknown set of values
+    Any,
+    // the empty set of values
+    Never,
+    // unknown type (no annotation)
+    // equivalent to Any, or to object in strict mode
+    Unknown,
+    // name is not bound to any value
+    Unbound,
+    // TODO protocols, callable types, overloads, generics, type vars
 }
 
 impl Type {
-    pub(crate) fn name(&self) -> &str {
+    pub(crate) fn name(&self) -> Option<&str> {
         match self {
-            Type::Class(inner) => inner.name(),
-            Type::Function(inner) => inner.name(),
+            Type::Instance(inner) => Some(inner.name()),
+            Type::Function(inner) => Some(inner.name()),
+            Type::Union(_) => None,
+            Type::Intersection(_) => None,
+            Type::Any => Some("Any"),
+            Type::Never => Some("Never"),
+            Type::Unknown => Some("Unknown"),
+            Type::Unbound => Some("Unbound"),
+        }
+    }
+
+    fn display(&self, f: &mut std::fmt::Formatter, env: &TypeEnvironment) -> std::fmt::Result {
+        if let Some(name) = self.name() {
+            write!(f, "{}", name)
```

Comment:
```suggestion
            f.write_str(name)
```

Using `write_str` bypasses the somewhat heavy format args compiler infrastructure